src/NuGet.Core/NuGet.Protocol.Core.Types/UserAgentStringBuilder.cs

```diff
@@ -0,0 +1,124 @@
+using NuGet.Versioning;
+
+namespace NuGet.Protocol.Core.Types
+{
+    public class UserAgentStringBuilder
+    {
+        public static readonly string DefaultNuGetClientName = "NuGet Client V3";
+
+        private const string UserAgentWithOSDescriptionAndVisualStudioSKUTemplate = "{0}/{1} ({2}, {3})";
+        private const string UserAgentWithOSDescriptionTemplate = "{0}/{1} ({2})";
+        private const string UserAgentTemplate = "{0}/{1}";
+
+        private readonly string _clientInfo;
+        private string _vsInfo;
+        private string _osInfo;
+
+        public UserAgentStringBuilder()
+            : this(DefaultNuGetClientName)
+        {
+        }
+
+        public UserAgentStringBuilder(string clientInfo)
+        {
+            _clientInfo = clientInfo;
+            NuGetClientVersion = GetNuGetVersion();
+        }
+
+        public NuGetVersion NuGetClientVersion { get; }
+
+        public UserAgentStringBuilder WithOSDescription(string osInfo)
+        {
+            _osInfo = osInfo;
+            return this;
+        }
+
+        public UserAgentStringBuilder WithVisualStudioSKU(string vsInfo)
+        {
+            _vsInfo = vsInfo;
+            return this;
+        }
+
+        public string Build()
+        {
+            var osDescription = _osInfo ?? GetOSVersion();
+
+            var clientInfo = _clientInfo;
+            if (NuGetTestMode.Enabled)
+            {
+                clientInfo = NuGetTestMode.NuGetTestClientName;
+            }
+            else if (string.IsNullOrEmpty(clientInfo))
+            {
+                clientInfo = DefaultNuGetClientName;
+            }
+
+            if (string.IsNullOrEmpty(osDescription))
+            {
+                return string.Format(
+                    CultureInfo.InvariantCulture,
+                    UserAgentTemplate,
+                    clientInfo,
+                    NuGetClientVersion);
+            }
+
+            if (string.IsNullOrEmpty(_vsInfo))
+            {
+                return string.Format(
+                    CultureInfo.InvariantCulture,
+                    UserAgentWithOSDescriptionTemplate,
+                    clientInfo,
+                    NuGetClientVersion,
+                    osDescription);
+            }
+            else
+            {
+                return string.Format(
+                    CultureInfo.InvariantCulture,
+                    UserAgentWithOSDescriptionAndVisualStudioSKUTemplate,
+                    _clientInfo,
+                    NuGetClientVersion, /* NuGet version */
+                    osDescription, /* OS version */
+                    _vsInfo);  /* VS SKU + version */
+            }
+        }
+
+        private static NuGetVersion GetNuGetVersion()
+        {
+            Version attr = null;
+
+#if !DNXCORE50
+            attr = typeof(Repository).Assembly.GetName().Version;
+#endif
+
+            if (attr == null)
+            {
+                return new NuGetVersion(3, 0, 0, 0);
```

Comment:
This should definitely be 3.4, it is pretty critical piece of statistics. if we can't do it in code (we should). We also need a test to validate it writes the correct version number.